src/components/ha-web-rtc-player.ts

```diff
@@ -119,98 +171,217 @@ class HaWebRtcPlayer extends LitElement {
     console.timeLog("WebRTC", "start createOffer", offerOptions);
 
     const offer: RTCSessionDescriptionInit =
-      await peerConnection.createOffer(offerOptions);
+      await this._peerConnection.createOffer(offerOptions);
+
+    if (!this._peerConnection) {
+      return;
+    }
 
     console.timeLog("WebRTC", "end createOffer", offer);
 
     console.timeLog("WebRTC", "start setLocalDescription");
 
-    await peerConnection.setLocalDescription(offer);
+    await this._peerConnection.setLocalDescription(offer);
+
+    if (!this._peerConnection || !this.entityid) {
+      return;
+    }
 
     console.timeLog("WebRTC", "end setLocalDescription");
 
-    console.timeLog("WebRTC", "start iceResolver");
-
-    let candidates = ""; // Build an Offer SDP string with ice candidates
-    const iceResolver = new Promise<void>((resolve) => {
-      peerConnection.addEventListener("icecandidate", (event) => {
-        if (!event.candidate?.candidate) {
-          resolve(); // Gathering complete
-          return;
-        }
-        console.timeLog("WebRTC", "iceResolver candidate", event.candidate);
-        candidates += `a=${event.candidate.candidate}\r\n`;
+    let candidates = "";
+
+    if (clientConfig.getCandidatesUpfront) {
+      await new Promise<void>((resolve) => {
+        this._peerConnection!.onicegatheringstatechange = (ev: Event) => {
+          const iceGatheringState = (ev.target as RTCPeerConnection)
+            .iceGatheringState;
+          if (iceGatheringState === "complete") {
+            resolve();
+          }
+
+          console.timeLog(
+            "WebRTC",
+            "Ice gathering state changed",
+            iceGatheringState
+          );
+        };
       });
-    });
-    await iceResolver;
 
-    console.timeLog("WebRTC", "end iceResolver", candidates);
+      if (!this._peerConnection || !this.entityid) {
+        return;
+      }
+    }
+
+    while (this._candidatesList.length) {
+      const candidate = this._candidatesList.pop();
+      if (candidate) {
+        candidates += `a=${this._candidatesList.pop()}\r\n`;
```

Comment:
You're popping twice here. Shouldn't you append the candidate?

```suggestion
      const candidate = this._candidatesList.pop();
      if (candidate) {
        candidates += `a=${candidate}\r\n`;
```